azure-devops/run-build.yml

```diff
@@ -19,12 +19,52 @@ jobs:
       displayName: 'Setup TMP Directory'
     - checkout: self
       clean: true
-      submodules: true
+      submodules: false
+    - task: PowerShell@2
+      displayName: 'Get submodule SHAs'
+      timeoutInMinutes: 1
+      inputs:
+        targetType: inline
+        script: |
+          cd $(Build.SourcesDirectory)
+          $pattern = "[^a-zA-Z0-9]"
+          $llvmSHA = git submodule status llvm-project | %{($_ -split "\s+")[0]}
+          $llvmSHA = $llvmSHA -replace $pattern,""

```

Comment:
While this works in the CI, I was confused by what it's doing. As you explained on Discord, this works because the submodule isn't initialized, so `git submodule status` prints a leading `-`. In a normal repository, a leading space will typically be printed, which interferes with this strategy of "split on whitespace, extract 0th element". I would like to recommend an alternate strategy that works in a single step: describe what we expect to see in a regex, and directly extract the SHA. Here is the regex that I propose, and examples where `git submodule status` prints `-` (for not initialized), a space (for initialized), or `+` (for initialized, but checked out to a different commit):

```
PS C:\Temp\DELME2\STL> $regexSubmoduleSHA = '^[ \-+]([0-9a-f]+) .*$'
PS C:\Temp\DELME2\STL> git submodule status llvm-project
-a668ad92d5e2161e07e1a435a19ea5072f52a989 llvm-project
PS C:\Temp\DELME2\STL> git submodule status llvm-project | %{$_ -replace $regexSubmoduleSHA, '$1'}
a668ad92d5e2161e07e1a435a19ea5072f52a989
```

```
PS C:\Temp\DELME\STL> $regexSubmoduleSHA = '^[ \-+]([0-9a-f]+) .*$'
PS C:\Temp\DELME\STL> git submodule status llvm-project
 a668ad92d5e2161e07e1a435a19ea5072f52a989 llvm-project (llvmorg-12-init-9073-ga668ad92d5e)
PS C:\Temp\DELME\STL> git submodule status llvm-project | %{$_ -replace $regexSubmoduleSHA, '$1'}
a668ad92d5e2161e07e1a435a19ea5072f52a989

PS C:\Temp\DELME\STL> cd llvm-project
PS C:\Temp\DELME\STL\llvm-project> git checkout HEAD~1
Previous HEAD position was a668ad92d5e [libc++][test] MSVC has no __PRETTY_FUNCTION__
HEAD is now at 7f82352a2c7 [libc++][test] Test nonconforming atomic_fetch_XXX overloads only on libc++
PS C:\Temp\DELME\STL\llvm-project> cd..

PS C:\Temp\DELME\STL> git submodule status llvm-project
+7f82352a2c766114bf1f2aa70f458da4e7304043 llvm-project (llvmorg-12-init-9072-g7f82352a2c7)
PS C:\Temp\DELME\STL> git submodule status llvm-project | %{$_ -replace $regexSubmoduleSHA, '$1'}
7f82352a2c766114bf1f2aa70f458da4e7304043
```

Similarly for `vcpkg` immediately below.

Reply:
Why not use the one-liner:
`$llvmSHA = git submodule status --cached llvm-project | %{($_ -split "\+|-|\s+")[1]}`

I also believe that adding `--cached` to `git submodule status` is desirable as we always want to end up with the submodule at the commit that is recorded in the repo.

Reply:
POSIX `cut -c 2-41` is laughing at us right now.

Reply:
POSIX is always laughing at me whenever I try to work on the CI. 😿

Reply:
`--cached` sounds good to me. (For the record, my examples above depicted plain `git submodule status llvm-project` only to show what the input to the regex was.)